src/SelfManage/Verify/FallbackVerificationUsingOpenSsl.php

```diff
@@ -0,0 +1,297 @@
+        private readonly string $githubApiBaseUrl,
+        private readonly HttpDownloader $httpDownloader,
+        private readonly AuthHelper $authHelper,
+    ) {
+    }
+
+    public function verify(ReleaseMetadata $releaseMetadata, BinaryFile $pharFilename, OutputInterface $output): void
+    {
+        $output->writeln(
+            'Falling back to basic verification. To use full verification, install the `gh` CLI tool.',
+            OutputInterface::VERBOSITY_VERBOSE,
+        );
+
+        $attestations = $this->downloadAttestations($releaseMetadata, $pharFilename);
+
+        foreach ($attestations as $attestationIndex => $attestation) {
+            /**
+             * Useful references. Whilst we don't do the full verification that
+             * `gh attestation verify` would (since we don't want to re-invent
+             * the wheel), we can do some basic check of the DSSE Envelope.
+             * We'll check the payload digest matches our expectation, and
+             * verify the signature with the certificate.
+             *
+             *  - https://github.com/cli/cli/blob/234d2effd545fb9d72ea77aa648caa499aecaa6e/pkg/cmd/attestation/verify/verify.go#L225-L256
+             *  - https://docs.sigstore.dev/logging/verify-release/
+             *  - https://github.com/secure-systems-lab/dsse/blob/master/protocol.md#protocol
+             */
+            $this->assertCertificateClaims($attestation);
+            $output->writeln('#' . $attestationIndex . ': Certificate facts verified.', OutputInterface::VERBOSITY_VERBOSE);
+
+            $this->assertDigestFromAttestationMatchesActual($pharFilename, $attestation);
+            $output->writeln('#' . $attestationIndex . ': Payload digest matches downloaded file.', OutputInterface::VERBOSITY_VERBOSE);
+
+            $this->verifyDsseEnvelopeSignature($releaseMetadata, $attestationIndex, $attestation);
+            $output->writeln('#' . $attestationIndex . ': DSSE payload signature verified with certificate.', OutputInterface::VERBOSITY_VERBOSE);
+        }
+
+        $output->writeln('<info>✅ Verified the new PIE version (using fallback verification)</info>');
+    }
+
+    private function assertCertificateClaims(Attestation $attestation): void
+    {
+        $attestationCertificateInfo = openssl_x509_parse($attestation->certificate);
+        Assert::isArray($attestationCertificateInfo['extensions']);
+
+        /**
+         * See {@link https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#136141572641--fulcio} for details
+         * on the Fulcio extension keys; note the values are DER-encoded strings; the ASN.1 tag is UTF8String (0x0C).
+         *
+         * First up, check the extension values are what we expect; these are hard-coded, as we don't expect them
+         * to change unless the namespace/repo name change, etc.
+         */
+        foreach (self::ATTESTATION_CERTIFICATE_EXPECTED_EXTENSION_VALUES as $extension => $expectedValue) {
```

Comment:
This should move to the bottom: It's a more logical verification order.

1. Verify that the certificate is a validly signed certificate in the first place.
2. Then verify that the certificate belongs to PIE.

Trying to extract data from an unverified certificate is questionable and can lead to confusing error messages).

Reply:
Makes sense, updated